mlflow/tracing/types/wrapper.py

```diff
@@ -187,44 +272,110 @@ def end(self):
+            "end_time": self.end_time_ns,
+            "status_code": self.status.status_code,
+            "status_message": self.status.description,
+            "attributes": dict(self._span.attributes),
+            "events": [asdict(event) for event in self.events],
+        }
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "Span":
+        request_id = json.loads(data["attributes"].get(SpanAttributeKey.REQUEST_ID))
+        if not request_id:
+            raise MlflowException(
+                "Request ID is missing in the loaded span data.", INVALID_PARAMETER_VALUE
+            )
 
-        :meta private:
-        """
-        return Span(
-            name=self._span.name,
-            context=SpanContext(
-                trace_id=self._trace_id,
-                span_id=self.span_id,
-            ),
-            parent_id=self.parent_id,
-            status_code=self.status.status_code.value,
-            status_message=self.status.description,
-            start_time=self.start_time_ns,
-            end_time=self.end_time_ns,
-            # Convert from MappingProxyType class to a simple dict
-            attributes=dict(self._span.attributes),
+        trace_id = decode_trace_id(data["context"]["trace_id"])
+        span_id = decode_span_id(data["context"]["span_id"])
+        parent_id = decode_span_id(data["parent_id"]) if data["parent_id"] else None
+
+        otel_span = OTelReadableSpan(
+            name=data["name"],
+            context=build_otel_context(trace_id, span_id),
+            parent=build_otel_context(trace_id, parent_id) if parent_id else None,
+            start_time=data["start_time"],
+            end_time=data["end_time"],
+            attributes=data["attributes"],
             events=[
-                SpanEvent(
-                    name=event.name,
-                    timestamp=event.timestamp,
-                    # Convert from OpenTelemetry's BoundedAttributes class to a simple dict
-                    attributes=dict(event.attributes),
+                OTelEvent(
+                    name=event["name"],
+                    timestamp=event["timestamp"],
+                    attributes=event["attributes"],
                 )
-                for event in self._span.events
+                for event in data["events"]
             ],
         )
+        return cls(otel_span, request_id)
+
+
+class _SpanAttributesRegistry:
+    """
+    An utility class to manage the span attributes.
+
+    In MLflow users can add arbitrary key-value pairs to the span attributes, however,
+    OpenTelemetry only allows a limited set of types to be stored in the attribute values.
+    Therefore, we serialize all values into JSON string before storing them in the span.
+    This class provides simple getter and setter methods to interact with the span attributes
+    without worrying about the serde process.
+    """
+
+    def __init__(self, otel_span: OTelSpan):
+        self._span = otel_span
```

Comment:
This seems redundant as we need to store otel_span twice. If the usage of this class is only to distinguish between non-cached and cached spans, can we use an attribute in Span maybe `is_cached_span` directly?

Reply:
This is passed as reference so we don't create copy of spans. Also we need to pass the span itself to access attribtues.

Reply:
Yea I see the span is needed to access the attributes, is this class necessary?

Reply:
If we don't have this class, how we dynamically enable/disable cache?

Reply:
Ah the deserialization of attributes does need cache...Adding additional condition inside `get` function seems more complex.
Another small Q: seems we can't avoid `set` function been called on _CachedSpanAttributesRegistry, do we want to override the `set` function to be not-supported?

Reply:
> do we want to override the set function to be not-supported?

That sounds a great idea!